### ISSUE TITLE
NVDA fails to activate a graphic button on a website

### ISSUE THREAD

**Issue by amirsol81** [nvaccess/nvda#7842]
In summary, NVDA detects and displays the "Download" graphic button in Firefox and Chrome, but fails to activate it. However, the button is missing in NVDA's rendering of the website when opened in Internet Explorer.

Steps to reproduce:
1. Go to http://www.wbur.org/hereandnow/2017/11/24/american-heritage-dictionary-new-words
2. Select the Download button, which is a graphic button, above the Play button.
3. You'll notice that NVDA does nothing.

Expected behavior:
Depending on the browser used, the Download dialog, either hidden or visible, should come up and the file should be downloaded, but nothing happens.

Test environment:
Tested with NVDA 2017.4 official release, Chrome 62, Firefox 57, and Internet Explorer 11, both on Windows 10 Pro 64-Bit and on Windows 7. Also tried activating the graphic button using both normal Quick Key navigation keys and the Elements List dialog to no avail.

Points worth mentioning:
1. With Internet Explorer 11, the button isn't detected or displayed at all in NVDA's virtual buffer. Instead NVDA displays an empty spot above the Play button.
2. JAWS doesn't have this issue at all and can activate the button/grab the file regardless of the browser used.

**Comment by Adriani90** [comment]
I can reproduce this. I have encountered this issue on some other Websites as well.

**Comment by feerrenrut** [comment]
I can also reproduce this in Chrome `63.0.3239.84 (Official Build) (64-bit)`
Steps I took:
1. went to the above website
1. Pressed `l` then tab to move the focus to the download button.
1. Pressed `enter`
 
I can confirm that without NVDA running the file downloads.

**Comment by feerrenrut** [comment]
Looking at this some more, after focusing the download button, doing the following will trigger the download:
1. Press `NVDA+ctrl+z` to open the NVDA python console
1. type `focus.doAction()`

I can also trigger the download by swapping to focus mode and pressing enter.

Technical:
1. `_activatePosition` is called in `class BrowseModeTreeInterceptor` , `obj` is `None`
1. `obj` gets set to `self.currentNVDAObject` (this does not match focus, see `devInfo` below)
1. `self._activateNVDAObject(obj)` is called, which calls the `class Gecko_ia2` override
1.`doAction` is called on `obj` which calls the `class IAccessible` override
1. No exception is thrown, and break is called exiting the loop in `def _activateNVDAObject` from `class Gecko_ia2`

`devInfo` for `obj` is:
```
["name: u'Download'", 'role: ROLE_GRAPHIC', 'states: ', 'isFocusable: False', 'hasFocus: False', 'Python object: <NVDAObjects.IAccessible.ia2Web.Ia2Web object at 0x0770BBF0>', "Python class mro: (<class 'NVDAObjects.IAccessible.ia2Web.Ia2Web'>, <class 'NVDAObjects.IAccessible.IAccessible'>, <class 'NVDAObjects.window.Window'>, <class 'NVDAObjects.NVDAObject'>, <class 'baseObject.ScriptableObject'>, <class 'baseObject.AutoPropertyObject'>, <type 'object'>)", 'description: None', 'location: (504, 548, 21, 20)', "value: u''", "appModule: <'appModuleHandler' (appName u'chrome', process ID 1680) at address 76942d0>", "appModule.productName: u'Google Chrome'", "appModule.productVersion: u'63.0.3239.84'", "TextInfo: <class 'NVDAObjects.IAccessible.IA2TextTextInfo'>", 'windowHandle: 1639912L', "windowClassName: u'Chrome_RenderWidgetHostHWND'", 'windowControlID: 378102112', 'windowStyle: 1445986304', 'windowThreadID: 884', "windowText: u'Chrome Legacy Window'", "displayText: u''", 'IAccessibleObject: <POINTER(IAccessible2) ptr=0x2986a8c at 766d580>', 'IAccessibleChildID: 0', 'IAccessible event parameters: windowHandle=1639912, objectID=-4, childID=-239084', "IAccessible accName: u'Download'", 'IAccessible accRole: ROLE_SYSTEM_GRAPHIC', 'IAccessible accState:  (0)', 'IAccessible accDescription: None', "IAccessible accValue: u''", 'IAccessible2 windowHandle: 1639912', 'IAccessible2 uniqueID: -239084', 'IAccessible2 role: ROLE_SYSTEM_GRAPHIC', 'IAccessible2 states: IA2_STATE_OPAQUE (1024)', "IAccessible2 attributes: u'display:inline;tag:svg;explicit-name:true;class:icon icon-download;'"]
```

devInfo for current focus is
```
["name: u'Download'", 'role: ROLE_BUTTON', 'states: STATE_FOCUSABLE, STATE_FOCUSED', 'isFocusable: True', 'hasFocus: True', 'Python object: <NVDAObjects.Dynamic_ButtonIa2WebIAccessible object at 0x0767B870>', "Python class mro: (<class 'NVDAObjects.Dynamic_ButtonIa2WebIAccessible'>, <class 'NVDAObjects.IAccessible.Button'>, <class 'NVDAObjects.IAccessible.ia2Web.Ia2Web'>, <class 'NVDAObjects.IAccessible.IAccessible'>, <class 'NVDAObjects.window.Window'>, <class 'NVDAObjects.NVDAObject'>, <class 'baseObject.ScriptableObject'>, <class 'baseObject.AutoPropertyObject'>, <type 'object'>)", "description: u'Download the audio'", 'location: (478, 542, 73, 26)', "value: u''", "appModule: <'appModuleHandler' (appName u'chrome', process ID 1680) at address 76942d0>", "appModule.productName: u'Google Chrome'", "appModule.productVersion: u'63.0.3239.84'", "TextInfo: <class 'NVDAObjects.IAccessible.IA2TextTextInfo'>", 'windowHandle: 1639912L', "windowClassName: u'Chrome_RenderWidgetHostHWND'", 'windowControlID: 378102112', 'windowStyle: 1445986304', 'windowThreadID: 884', "windowText: u'Chrome Legacy Window'", "displayText: u''", 'IAccessibleObject: <POINTER(IAccessible2) ptr=0x2986a44 at 766d7b0>', 'IAccessibleChildID: 0', 'IAccessible event parameters: windowHandle=1639912, objectID=-4, childID=-239077', "IAccessible accName: u'Download'", 'IAccessible accRole: ROLE_SYSTEM_PUSHBUTTON', 'IAccessible accState: STATE_SYSTEM_FOCUSED, STATE_SYSTEM_FOCUSABLE, STATE_SYSTEM_VALID (1048580)', "IAccessible accDescription: u'Download the audio'", "IAccessible accValue: u''", 'IAccessible2 windowHandle: 1639912', 'IAccessible2 uniqueID: -239077', 'IAccessible2 role: ROLE_SYSTEM_PUSHBUTTON', 'IAccessible2 states: IA2_STATE_OPAQUE (1024)', "IAccessible2 attributes: u'display:block;tag:a;xml-roles:button;class:article-audio-dl;'"]
```

**Comment by amirsol81** [comment]
Thanks for the technical analysis! Any chance of coming up with a fix?
Moreover, the button isn't detected in IE11's buffer at all, so I'm
wondering how the fix can be extended to Internet Explorer 11.

**Comment by feerrenrut** [comment]
Despite some time spent on investigation the exact cause is not clear. While I understand that this is an annoying issue, there is a work around so I am going to give this a priority 3 to fix.

**Comment by amirsol81** [comment]
Thanks for looking into this.
Of course, I can't find the work-around... I might have missed
something here, though.

**Comment by feerrenrut** [comment]
The simplest work around is to swap to focus mode `NVDA+space` and then press enter.

**Comment by amirsol81** [comment]
Thanks -- great tip!

**Comment by Adriani90** [comment]
Opening this website in IE11 and switching to focus mode, then navigating with tab and shift+tab, makes NVDA unusable after a while and the log is full of errors. This is one of them:

```
IO - inputCore.InputManager.executeGesture (12:06:48.913):
Input: kb(laptop):downArrow
ERROR - scriptHandler.executeScript (12:06:49.023):
error executing script: <bound method MSHTML.script_moveByLine_forward of <virtualBuffers.MSHTML.MSHTML object at 0x08497330>> with gesture u'Pfeiltaste nach unten'
Traceback (most recent call last):
  File "scriptHandler.pyc", line 187, in executeScript
  File "cursorManager.pyc", line 254, in script_moveByLine_forward
  File "cursorManager.pyc", line 185, in _caretMovementScriptHelper
  File "browseMode.pyc", line 1245, in _set_selection
  File "browseMode.pyc", line 1260, in _shouldSetFocusToObj
  File "baseObject.pyc", line 47, in __get__
  File "baseObject.pyc", line 147, in _getPropertyViaCache
  File "NVDAObjects\IAccessible\MSHTML.pyc", line 944, in _get_isFocusable
  File "comtypes\client\lazybind.pyc", line 226, in __call__
  File "comtypes\automation.pyc", line 729, in _invoke
COMError: (-2147352319, None, (None, None, None, 0, None))
```

Many parts of the website including the download button are not reported by NVDA in IE due to these errors.

**Comment by Adriani90** [comment]
In Firefox the player opens and the audio is played. In Chrome nothing happens when pressing enter on the download button.

**Comment by Adriani90** [comment]
Now in Chrome 81 also the audio player opens when pressing on the download button. I think this is a bad web authoring here. But I am not sure.
cc: @MarcoZehe and @aleventhal maybe you have some thoughts on this?

**Comment by MarcoZehe** [comment]
I can't reproduce.

**Comment by Adriani90** [comment]
@MarcoZehe did you try to compare Firefox and Chrome for example? In Firefox the download button is not displayed at all in the virtual buffer.

**Comment by MarcoZehe** [comment]
I *did* see the download button. That's why I said that I cannot reproduce.

**Comment by Adriani90** [comment]
This is an web authoring issue here. Testing with NVDA 2023.1 RC1 and Chrome 111 or Firefox 110, nothing triggers a download on this button. In fact, even when turning off NVDA and clicking on the button, the file is not download, but is played instead.

It seems the authon put a link button with a child SVG which should trigger the download but it doesn't work.

Cosing as not planed, since it is not related to NVDA.
